File: js/loadShedder.js
/*
 * Automatic load-shedding routine (design doc §3.3 priority strategy, §6).
 * Mirrors LoadShedder.java: when total bus current stays above the budget ceiling
 * for a short window, reduce the smart-current limit of the least-critical
 * subsystem first, then the next, etc. The drivetrain is never auto-shed.
 * Limits restore once draw drops below a hysteresis threshold.
 */
window.PM = window.PM || {};

PM.LoadShedder = class {
  constructor() {
    this.shedAbove = PM.Budget.TOTAL_CURRENT_BUDGET_AMPS;   // start shedding above 90A
    this.restoreBelow = PM.Budget.TOTAL_CURRENT_CAUTION_AMPS;// restore below 70A (hysteresis)
    this.sustainedLoops = 5;                                 // ~100ms @ 20ms
    this.reductionFraction = 0.5;                            // shed to 50% of normal limit
    this.enabled = true;

    this._overBudgetLoops = 0;
    this.shed = new Set(); // keys of currently-shed subsystems
  }

  reset() {
    this._overBudgetLoops = 0;
    this.shed.clear();
  }

  /*
   * Advance the state machine one loop. `subsystems` is the array of subsystem
   * defs sorted by priority. Returns the set of shed keys.
   */
  update(totalCurrent, subsystems) {
    if (!this.enabled) { this.shed.clear(); return this.shed; }

    if (totalCurrent > this.shedAbove) this._overBudgetLoops++;
    else this._overBudgetLoops = 0;

    if (this._overBudgetLoops >= this.sustainedLoops) {
      this._shedNextLeastCritical(subsystems);
    } else if (totalCurrent < this.restoreBelow && this.shed.size > 0) {
      this.shed.clear();
      this._overBudgetLoops = 0;
    }
    return this.shed;
  }

  _shedNextLeastCritical(subsystems) {
    // Walk from least critical (highest priority number) to most critical.
    const ordered = [...subsystems].sort((a, b) => b.priority - a.priority);
    for (const sub of ordered) {
      if (!sub.sheddable) continue;        // never shed the drivetrain
      if (!this.shed.has(sub.key)) {
        this.shed.add(sub.key);
        return sub;                        // shed one per event
      }
    }
    return null; // everything sheddable already shed
  }

  isShedding() { return this.shed.size > 0; }

  // Effective smart-current limit for a subsystem given shed state.
  effectiveLimit(sub) {
    return this.shed.has(sub.key) ? sub.smart * this.reductionFraction : sub.smart;
  }
};

File: robot/src/main/java/frc/robot/power/LoadShedder.java
package frc.robot.power;

import edu.wpi.first.wpilibj.DriverStation;
import frc.robot.Constants.PowerBudget;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Automatic load-shedding routine (design doc &sect;3.3 priority strategy, &sect;6 future improvement).
 *
 * <p>When total bus current climbs above the budget ceiling, this temporarily reduces the
 * current limits of non-critical subsystems, sheddng the LEAST critical first (highest
 * {@link PowerSubsystem#shedPriority()}). The drivetrain (priority 0) is shed last. Limits are
 * restored once total draw recovers below a hysteresis threshold, so we don't oscillate.
 */
public class LoadShedder {

  /** Start shedding when sustained total current exceeds the budget ceiling. */
  private static final double SHED_ABOVE_AMPS = PowerBudget.TOTAL_CURRENT_BUDGET_AMPS;

  /** Restore once we drop back under this (hysteresis band prevents chatter). */
  private static final double RESTORE_BELOW_AMPS = PowerBudget.TOTAL_CURRENT_CAUTION_AMPS;

  /**
   * Total current must stay above the shed threshold for this many consecutive loops before we
   * act, so a brief acceleration+shot peak doesn't trigger shedding. 5 loops @ 20 ms = 100 ms.
   */
  private static final int SUSTAINED_LOOPS_TO_SHED = 5;

  private final List<CurrentLimited> registered = new ArrayList<>();
  private final List<CurrentLimited> currentlyShed = new ArrayList<>();

  private int loopsOverBudget = 0;

  /** Register a mechanism that is willing to be throttled. Order does not matter. */
  public void register(CurrentLimited mechanism) {
    registered.add(mechanism);
    // Keep sorted most-critical-first; we iterate in reverse to shed least-critical first.
    registered.sort(Comparator.comparingInt(m -> m.powerSubsystem().shedPriority()));
  }

  /**
   * Drive the shed/restore state machine. Call once per loop from the monitor with the latest
   * total bus current.
   *
   * @return true if any subsystem is currently shed
   */
  public boolean update(double totalCurrentAmps) {
    if (totalCurrentAmps > SHED_ABOVE_AMPS) {
      loopsOverBudget++;
    } else {
      loopsOverBudget = 0;
    }

    if (loopsOverBudget >= SUSTAINED_LOOPS_TO_SHED) {
      shedNextLeastCritical();
    } else if (totalCurrentAmps < RESTORE_BELOW_AMPS && !currentlyShed.isEmpty()) {
      restoreAll();
    }

    return !currentlyShed.isEmpty();
  }

  /** Shed one more subsystem, the least critical not-yet-shed one. */
  private void shedNextLeastCritical() {
    // registered is sorted most-critical-first; walk from the end to find the least critical
    // subsystem that isn't shed yet.
    for (int i = registered.size() - 1; i >= 0; i--) {
      CurrentLimited m = registered.get(i);
      if (m.powerSubsystem() == PowerSubsystem.DRIVETRAIN) {
        continue; // never auto-shed the drivetrain
      }
      if (!currentlyShed.contains(m)) {
        m.applyReducedLimit();
        currentlyShed.add(m);
        DriverStation.reportWarning(
            "Load-shedding: reduced current limit on " + m.powerSubsystem().label(), false);
        return;
      }
    }
    // If we get here every sheddable subsystem is already shed; nothing more we can do.
  }

  /** Restore every shed subsystem to its normal limit. */
  public void restoreAll() {
    for (CurrentLimited m : currentlyShed) {
      m.restoreNormalLimit();
    }
    if (!currentlyShed.isEmpty()) {
      DriverStation.reportWarning("Load-shedding cleared: current limits restored", false);
    }
    currentlyShed.clear();
    loopsOverBudget = 0;
  }

  /** Names of subsystems currently shed, for the dashboard. Empty string when none. */
  public String shedStatus() {
    if (currentlyShed.isEmpty()) {
      return "";
    }
    StringBuilder sb = new StringBuilder();
    for (CurrentLimited m : currentlyShed) {
      if (sb.length() > 0) {
        sb.append(", ");
      }
      sb.append(m.powerSubsystem().label());
    }
    return sb.toString();
  }

  public boolean isShedding() {
    return !currentlyShed.isEmpty();
  }
}
